File: doublecheckmeplease/other/get_new_dsts.rb
#!/usr/bin/env ruby

# frozen_string_literal: true

require 'json'
require 'open-uri'
require 'rest_client'
require 'dotenv'

Dotenv.load('../.env')

def search
  response = RestClient.get "#{ENV['zendesk_url']}/api/v2/search.json?query=need%20dst%20status:open%20status:pending%20status:new%20order_by:updated_at%20sort:desc%20has_attachment:true", Authorization: "Bearer #{ENV['zendesk_oauth']}"
  JSON.parse(response.body)
end

def filter_search(search)
  tickets_with_dsts_available = []
  search['results'].each do |ticket|
    subject = ticket['subject'].downcase.gsub(/,/, '').split(' ')
    subject.each { |s| tickets_with_dsts_available.push ticket['id'] if s.match?(/d...../) }
  end
  tickets_with_dsts_available
end

def download_dsts(ticket_number)
  response = RestClient.get "#{ENV['zendesk_url']}/api/v2/tickets/#{ticket_number}/comments.json", content_type: :json, Authorization: "Bearer #{ENV['zendesk_oauth']}"
  comments = JSON.parse(response.body)
  comments['comments'].each do |comment|
    comment['attachments'].each do |attachment|
      open(attachment['file_name'], 'wb') do |file|
        file << open(attachment['content_url']).read
      end
    end
  end
end

def comment_and_close(ticket_number)
  jdata = '{ "ticket": { "status": "solved", "comment": { "body": "Automatically processed", "author_id": 380031530892, "public": false } } }'
  RestClient.put "#{ENV['zendesk_url']}/api/v2/tickets/#{ticket_number}.json", jdata, content_type: :json, Authorization: "Bearer #{ENV['zendesk_oauth']}"
end

filter_search(search).each do |ticket_number|
  download_dsts ticket_number
  comment_and_close ticket_number
end

File: scratch/projects/zendesk_attachment_grabber/get_new_dsts.py
"""Check for new tickets with DSTs available, download the DSTs, and close the ticket."""

import json
import re
from pathlib import Path

import httpx
from pydantic_settings import BaseSettings, SettingsConfigDict


class EnvConfig(BaseSettings):
    """Our default configuration for models that should load from .env files."""

    model_config = SettingsConfigDict(
        env_file=".env",
        env_file_encoding="utf-8",
        env_nested_delimiter="__",
        extra="ignore",
    )


class _Zendesk(EnvConfig, env_prefix="zendesk_"):
    """UPS auth."""

    base_url: str = ""
    username: str = ""
    api_token: str = ""


Zendesk = _Zendesk()


def search(client: httpx.Client) -> list[int]:
    """Find open tickets titled "need dst" with attachments and return the ticket numbers."""
    params = {
        "query": "need dst status:open status:pending status:new order_by:updated_at sort:desc has_attachment:true",
    }
    response = client.get("/api/v2/search.json", params=params)

    tickets_with_dsts_available = []
    for ticket in response.json()["results"]:
        subject_pieces = ticket["subject"].lower().replace(",", "").split(" ")
        for piece in subject_pieces:
            if re.match(r"d[\w\d]{5}", piece):
                tickets_with_dsts_available.append(ticket["id"])  # noqa: PERF401 -- false positive, not possible
    return tickets_with_dsts_available


def download_dsts(client: httpx.Client, ticket_number: int) -> None:
    response = client.get(f"/api/v2/tickets/{ticket_number}/comments.json")
    comments = response.json()
    for comment in comments["comments"]:
        for attachment in comment["attachments"]:
            # see if this matches the base URL
            attachment_bytes = client.get(attachment["content_url"], auth=client.auth).content
            Path(attachment["file_name"]).write_bytes(attachment_bytes)


def comment_and_close(client: httpx.Client, ticket_number: int) -> None:
    data = {
        "ticket": {
            "status": "solved",
            "comment": {
                "body": "Automatically processed",
                "author_id": 380031530892,
                "public": False,
            },
        },
    }
    client.put(f"/api/v2/tickets/{ticket_number}.json", data=json.dumps(data))


def run() -> None:
    auth = httpx.BasicAuth(username=Zendesk.username, password=Zendesk.api_token)
    client = httpx.Client(base_url=Zendesk.base_url, auth=auth, headers={"Content-Type": "application/json"})

    for ticket_number in search(client):
        download_dsts(ticket_number)
        # comment_and_close(ticket_number)


if __name__ == "__main__":
    run()
